<a href="https://colab.research.google.com/github/shakurovas/pt-ms_gb/blob/master/homework6_pt%26ms.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import numpy as np
import pandas as pd
import scipy.stats as st

**Задание 1:** Даны значения величины заработной платы заемщиков банка (zp) и значения их поведенческого кредитного скоринга (ks):
zp = [35, 45, 190, 200, 40, 70, 54, 150, 120, 110],
ks = [401, 574, 874, 919, 459, 739, 653, 902, 746, 832].
Найдите ковариацию этих двух величин с помощью элементарных действий, а затем с помощью функции cov из numpy.
Полученные значения должны быть равны.
Найдите коэффициент корреляции Пирсона с помощью ковариации и среднеквадратичных отклонений двух признаков,
а затем с использованием функций из библиотек numpy и pandas.

**Решение:**

In [ ]:
zp = np.array([35, 45, 190, 200, 40, 70, 54, 150, 120, 110])
ks = np.array([401, 574, 874, 919, 459, 739, 653, 902, 746, 832])

Ковариация через встроенную функцию:

In [ ]:
cov_offset = np.cov(zp, ks, ddof=0)[0][1]
cov_without_offset = np.cov(zp, ks, ddof=1)[0][1]
print(cov_offset)
print(cov_without_offset)

9157.84
10175.377777777776


Ковариация через расчёты:

In [ ]:
cov = np.mean(zp * ks) - np.mean(zp) * np.mean(ks)
print(cov)

9157.839999999997


Коэффициент корреляции Пирсона через встроенные функции:

**Numpy:**

In [ ]:
np.corrcoef(zp, ks)

array([[1.        , 0.88749009],
       [0.88749009, 1.        ]])

**Pandas:**

In [ ]:
df = pd.DataFrame(list(zip(zp, ks)))
corr_coef = df.corr(method='pearson')
print(corr_coef)

         0        1
0  1.00000  0.88749
1  0.88749  1.00000


Коэффициент корреляции Пирсона через расчёты:

In [ ]:
pearson_coef_offset =  cov_offset / np.std(zp, ddof=0) / np.std(ks, ddof=0)
print(pearson_coef_offset)

0.8874900920739162


In [ ]:
pearson_coef_without_offset =  cov_without_offset / np.std(zp, ddof=1) / np.std(ks, ddof=1)
print(pearson_coef_without_offset)

0.8874900920739159


**Задание 2:** Измерены значения IQ выборки студентов,
обучающихся в местных технических вузах:
131, 125, 115, 122, 131, 115, 107, 99, 125, 111.
Известно, что в генеральной совокупности IQ распределен нормально.
Найдите доверительный интервал для математического ожидания с надежностью 0.95.

In [ ]:
iq_list = np.array([131, 125, 115, 122, 131, 115, 107, 99, 125, 111])

In [ ]:
iq_list_mean = np.mean(iq_list)
print(iq_list_mean)

118.1


In [ ]:
iq_list_sigma = np.std(iq_list, ddof=1)
print(iq_list_sigma)

10.54566788359614


In [ ]:
accuracy_2 = 0.95
n_2 = iq_list.shape[0]
print(n_2)

10


Т. к. дисперсия генеральной совокупности нам неизвестна, используем критерий Стьюдента:

In [ ]:
t = abs(st.t.ppf((1 - accuracy_2) / 2, n - 1)) # то, что критерий Стьюдента получается отрицательным, говорит о том, что среднее выборочное меньше мат. ожидания изначальной выборки
print(t)

2.2621571627409915


In [ ]:
conf_int_left_2 = iq_list_mean - t * iq_list_sigma / np.sqrt(n_2)
print(conf_int_left_2)

110.55608365158724


In [ ]:
conf_int_right_2 = iq_list_mean + t * iq_list_sigma / np.sqrt(n_2)
print(conf_int_right_2)

125.64391634841274


или:

In [ ]:
st.t.interval(0.95, 9, loc=iq_list_mean, scale=st.sem(iq_list))

(110.55608365158724, 125.64391634841274)

Итак, с надежностью 0.95 истинное значение измеряемой величины заключено в доверительном интервале 110.556 <= X <= 125.644

**Задание 3:** Известно, что рост футболистов в сборной распределен нормально
с дисперсией генеральной совокупности, равной 25 кв.см. Объем выборки равен 27,
среднее выборочное составляет 174.2. Найдите доверительный интервал для математического
ожидания с надежностью 0.95.

**Решение:**

In [ ]:
var = 25
n_3 = 27
mean = 174.2
accuracy_3 = 0.95

Т. к. дисперсия генеральной совокупности нам известна (var=25), то используем z-критерий:

In [ ]:
z = st.norm.isf((1 - accuracy_3) / 2)
print(z)

1.959963984540054


In [ ]:
conf_int_left_3 = mean - z / np.sqrt(var / n_3)
print(conf_int_left_3)

172.1631456786629


In [ ]:
conf_int_right_3 = mean + z / np.sqrt(var / n_3)
print(conf_int_right_3)

176.2368543213371


Итак, с надежностью 0,95 истинное значение измеряемой величины заключено в доверительном интервале 172.163 <= X <= 176.237